src/Hyperrectangle.jl

```diff
@@ -34,56 +55,34 @@ end
 Hyperrectangle(center::Vector{N}, radius::Vector{N}) where {N<:Real} =
     Hyperrectangle{N}(center, radius)
 
-"""
-    Hyperrectangle(;kwargs...)
-
-Construct a hyperrectangle from keyword arguments.
-
-### Input
-
-- `kwargs` -- keyword arguments; two combinations are allowed:
-  1. `center`, `radius` -- vectors
-  2. `high`, `low`      -- vectors (if both `center` and `radius` are also
-                           defined, those are chosen instead)
-
-### Output
-
-A hyperrectangle.
-
-### Examples
-
-The following three constructions are equivalent:
-
-```jldoctest
-julia> c = ones(2);
-
-julia> r = [0.1, 0.2];
-
-julia> l = [0.9, 0.8];
-
-julia> h = [1.1, 1.2];
-
-julia> H1 = Hyperrectangle(c, r)
-LazySets.Hyperrectangle{Float64}([1.0, 1.0], [0.1, 0.2])
-julia> H2 = Hyperrectangle(center=c, radius=r)
-LazySets.Hyperrectangle{Float64}([1.0, 1.0], [0.1, 0.2])
-julia> H3 = Hyperrectangle(low=l, high=h)
-LazySets.Hyperrectangle{Float64}([1.0, 1.0], [0.1, 0.2])
-```
-"""
+# constructor from keyword arguments (lower and upper bounds)
+if VERSION < v"0.7-"
+@eval begin
 function Hyperrectangle(;kwargs...)
     dict = Dict{Symbol, Any}(kwargs)
-    if haskey(dict, :center) && haskey(dict, :radius)
-        return Hyperrectangle(dict[:center], dict[:radius])
-    elseif haskey(dict, :high) && haskey(dict, :low)
+    if haskey(dict, :high) && haskey(dict, :low)
         # compute center and radius from high and low vectors
-        center = (dict[:high] .+ dict[:low]) ./ 2
-        radius = abs.(dict[:high] .- center)
+        high = dict[:high]
+        center = (high .+ dict[:low]) ./ 2
+        radius = abs.(high .- center)
         return Hyperrectangle(center, radius)
     end
-    throw(ArgumentError("invalid arguments for Hyperrectangle: Use either " *
-        "'center' and 'radius' or 'high' and 'low'."))
+    throw(ArgumentError("invalid arguments for Hyperrectangle: " *
+        "use 'high' and 'low'"))
+end
+end # @eval
+else # if VERSION < v"0.7-"
+@eval begin
+function Hyperrectangle(;
+                        high::AbstractVector{N},
+                        low::AbstractVector{N}) where {N<:Real}
+    # compute center and radius from high and low vectors
+    center = (high .+ low) ./ 2
+    radius = abs.(high .- center)
+    return Hyperrectangle(center, radius)
```

Comment:
does it "help" the compiler to write `Hyperrectangle{N}(center, radius)` here?

Reply:
I am not completely sure, but `Hyperrectangle(center, radius)` falls back to `Hyperrectangle{N}(center, radius)`, so I hope that this step is skipped by the compiler.

Reply:
you're right. both versions compile to the same code

```julia
julia> @code_warntype Hyperrectangle2(high=rand(2), low=rand(2))
Body::Hyperrectangle{Float64}
 1 ─ %1  = (Base.getfield)(#temp#, :high)::Array{Float64,1}                                                                                        │╻     getindex
 │   %2  = (Base.getfield)(#temp#, :low)::Array{Float64,1}                                                                                         ││    
 │   %3  = (Base.slt_int)(0, 1)::Bool                                                                                                              ││╻╷╷╷  iterate
 └──       goto #3 if not %3                                                                                                                       │││┃│    iterate
 2 ─       goto #4                                                                                                                                 ││││┃     iterate
 3 ─       invoke Base.getindex(()::Tuple{}, 1::Int64)                                                                                             │││││ 
 └──       $(Expr(:unreachable))                                                                                                                   │││││ 
 4 ┄       goto #5                                                                                                                                 ││││  
 5 ─       goto #6                                                                                                                                 ││╻     iterate
 6 ─       goto #7                                                                                                                                 ││    
 7 ─       nothing                                                                                                                                 │     
 │   %12 = invoke Main.:(#Hyperrectangle2#4)(%1::Array{Float64,1}, %2::Array{Float64,1}, _3::Function)::Hyperrectangle{Float64}                    │     
 └──       return %12  
```